### PR TITLE
Support GSM cellular modem with PPP

### PR DESCRIPTION
**Pull request by pfl**
This patch set enables support for standard GSM AT command based cellular modems providing PPP connectivity. AT commands are sent to the modem, and after the modem has connected to the cellular network, the Zephyr PPP software stack is started and IPCP negotiated. Once an IPv4 address has been received, the PPP network interface is brought up.

The setup has been tested with Reel Board where the modem was connected via external UART_1 on the connector board. A simila setup was attempted for FRDM board, for that setup a DTS overlay file is provided. A FONA 808 modem was used for testing.

### CODE REVIEW

File: drivers/modem/Kconfig.gsm [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	 the range 0-99.
+
+config MODEM_GSM_APN
+       string "Access Point Name"
+       default "internet"
+       help
+         Specify Access Point Name, i.e. the name to identify Internet or
+	 IP GPRS/3g/4G/etc. cellular data context.
+
+endif
```
Comment (tbursztyka) [inline drivers/modem/Kconfig.gsm]: you'll need an line break here or make menuconfig will blow out

File: samples/net/gsm_modem/src/main.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	struct device *uart_dev = device_get_binding(CONFIG_MODEM_GSM_UART_NAME);
+
+	LOG_INF("%s: booted board '%s' APN '%s' UART '%s' device %p",
+                __FUNCTION__, CONFIG_BOARD, CONFIG_MODEM_GSM_APN,
+                CONFIG_MODEM_GSM_UART_NAME, uart_dev);
+
+	net_mgmt_init_event_callback(&mgmt_cb, event_handler,
+				     NET_EVENT_L4_CONNECTED |
+                                     NET_EVENT_L4_DISCONNECTED);
+        net_mgmt_add_event_callback(&mgmt_cb);
```
Comment (tbursztyka) [inline samples/net/gsm_modem/src/main.c]: useless spaces after the tab?

**Review by tbursztyka — APPROVED** [review]
besides 2 minor commets, lgtm

File: drivers/modem/Kconfig.gsm (hunk 1 above)
Comment (pfl) [reply drivers/modem/Kconfig.gsm]: Fixed.

File: samples/net/gsm_modem/src/main.c (hunk 2 above)
Comment (pfl) [reply samples/net/gsm_modem/src/main.c]: Yes.

Comment (tbursztyka) [reply samples/net/gsm_modem/src/main.c]: looks like checkpatch found a ton of such issue

File: drivers/modem/Kconfig.gsm [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	help
+	  The GSM modem is initialized in POST_KERNEL using priority in
+	  the range 0-99.
+
+config MODEM_GSM_APN
+	string "Access Point Name"
+	default "internet"
+	help
+	  Specify Access Point Name, i.e. the name to identify Internet or
+	  IP GPRS/3g/4G/etc. cellular data context.
```
Comment (pfalcon) [inline drivers/modem/Kconfig.gsm]: Any reason there's "3g", but then "4G"?

File: samples/net/gsm_modem/src/main.c (hunk 2 above)
Comment (pfl) [reply samples/net/gsm_modem/src/main.c]: Fixed a ton of them, that's what one gets by renaming and copy-pasting.

File: drivers/modem/Kconfig.gsm (hunk 3 above)
Comment (pfl) [reply drivers/modem/Kconfig.gsm]: No.

Comment (jukkar) [reply drivers/modem/Kconfig.gsm]: Might be better to remove "3G/4G" here until we have a modem that supports these and we have tested them.

File: drivers/modem/gsm_ppp.c [hunk 4]
```diff
@@ -46,13 +56,39 @@ struct k_thread gsm_rx_thread;
...
+					   &bytes);
+		if (r < 0 || bytes == 0) {
+			continue;
+		}
+
+		recv_buf_offset += bytes;
+
+		gsm.ppp_recv_buf = gsm.ppp_recv_cb(gsm.ppp_recv_buf,
+						   &recv_buf_offset);
+
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c]: extra empty line can be removed

File: samples/net/gsm_modem/prj.conf [hunk 5]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (jukkar) [inline samples/net/gsm_modem/prj.conf]: Better to put reel board specific stuff into `board/reel_board.conf` like we have in other samples.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: drivers/modem/gsm_ppp.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	k_thread_create(&gsm_rx_thread, gsm_rx_stack,
+			K_THREAD_STACK_SIZEOF(gsm_rx_stack),
+			(k_thread_entry_t) gsm_rx,
+			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+
+	k_delayed_work_init(&gsm.gsm_configure_work, gsm_configure);
+
+	(void) k_delayed_work_submit(&gsm.gsm_configure_work, 0);
```
[the remark this answers is not in the document]
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: Done.

File: drivers/modem/gsm_ppp.c (hunk 4 above)
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: Removed.

File: drivers/modem/gsm_ppp.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+			if (r < 0) {
+				LOG_DBG("modem not ready %d", r);
+			} else {
+				LOG_DBG("connect with modem %d", r);
+				break;
+			}
+			k_sleep(2000);
+		}
+
+		k_sleep(2000);
```
[the remark this answers is not in the document]
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: These k_sleeps are unnecessary, removed.

File: drivers/modem/Kconfig.gsm (hunk 3 above)
Comment (pfl) [reply drivers/modem/Kconfig.gsm]: Removed wording.

File: samples/net/gsm_modem/prj.conf (hunk 5 above)
Comment (pfl) [reply samples/net/gsm_modem/prj.conf]: Let's remove FRDM board clutter, but leave Reel board stuff. Right now we haven't really enabled it elsewhere and without an existing UART one can't really run the sample anyway.

Comment (pfl) [reply samples/net/gsm_modem/prj.conf]: Reel Board things are now in boards/

File: samples/net/gsm_modem/prj.conf [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+# IPv4 enables PPP IPCP support
+CONFIG_NET_IPV4=y
+CONFIG_NET_IPV6=n
+
+# Network management events
+CONFIG_NET_CONNECTION_MANAGER=y
+
+# Log buffers, modem and PPP
+CONFIG_LOG_BUFFER_SIZE=16384
```
Comment (jukkar) [inline samples/net/gsm_modem/prj.conf]: We need to modify the log settings here. I suggest that this is changed to (by default the debug logging should not be enabled). Also \n needs to be added to the end of the file.
```
CONFIG_LOG=y
CONFIG_NET_LOG=y
#CONFIG_LOG_BUFFER_SIZE=16384
#CONFIG_LOG_STRDUP_BUF_COUNT=200
#CONFIG_MODEM_LOG_LEVEL_DBG=y
#CONFIG_NET_PPP_LOG_LEVEL_DBG=y
#CONFIG_NET_L2_PPP_LOG_LEVEL_DBG=y
#CONFIG_NET_MGMT_EVENT_LOG_LEVEL_DBG=y
#CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL_DBG=y 
```

File: samples/net/gsm_modem/src/main.c [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+			  u32_t mgmt_event, struct net_if *iface)
+{
+	if ((mgmt_event & (NET_EVENT_L4_CONNECTED
+			   | NET_EVENT_L4_DISCONNECTED)) != mgmt_event) {
+		return;
+	}
+
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
+		LOG_INF("Network connected");
+	return;
```
Comment (jukkar) [inline samples/net/gsm_modem/src/main.c]: Indentation

File: drivers/modem/gsm_ppp.c [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+			}
+		}
+
+		r = modem_cmd_handler_setup_cmds(&gsm->context.iface,
+						 &gsm->context.cmd_handler,
+						 setup_cmds,
+						 ARRAY_SIZE(setup_cmds),
+						 &gsm->sem_response,
+						 GSM_CMD_SETUP_TIMEOUT);
+
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c]: This empty line can be removed

File: drivers/modem/gsm_ppp.c [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	struct k_sem sem_response;
+
+	struct modem_iface_uart_data gsm_data;
+	struct k_delayed_work gsm_configure_work;
+	char gsm_isr_buf[PPP_MRU];
+	char gsm_rx_rb_buf[PPP_MRU * 3];
+} gsm;
+
+NET_BUF_POOL_DEFINE(gsm_recv_pool, GSM_RECV_MAX_BUF, GSM_RECV_BUF_SIZE,
+		    0, NULL);
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c]: Indentation might be slightly off here (at least the diff looks a bit off)

File: drivers/modem/gsm_ppp.c [hunk 12]
```diff
@@ -46,13 +56,38 @@ struct k_thread gsm_rx_thread;
...
 	while (true) {
 		k_sem_take(&gsm.gsm_data.rx_sem, K_FOREVER);
-		gsm.context.cmd_handler.process(&gsm.context.cmd_handler,
-						&gsm.context.iface);
-		k_yield();
+
+		if (gsm.setup_done == false) {
+			gsm.context.cmd_handler.process(&gsm.context.cmd_handler,
+							&gsm.context.iface);
+
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c]: Empty line here could be removed

File: drivers/modem/gsm_ppp.c [hunk 13]
```diff
@@ -124,8 +159,14 @@ static void gsm_configure(struct k_work *work)
 
 		if (r < 0) {
 			LOG_DBG("modem setup returned %d, retrying...", r);
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c]: As the strings are almost the same, we could save some flash space by writing these like this
```
		if (r < 0) {
			LOG_DBG("modem setup returned %d, %s", r, "retrying...");
		} else {
			LOG_DBG("modem setup returned %d, %s", r, "enable PPP");
			break;
		}
```

**Review by jukkar — COMMENTED** [review]
Looks good, found couple of minor nits.

File: drivers/modem/gsm_ppp.c (hunk 11 above)
Comment (jukkar) [reply drivers/modem/gsm_ppp.c]: Verified with the source file, no issue in this line with indentation.

File: drivers/modem/gsm_ppp.c [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+	SETUP_CMD_NOHANDLE("AT+CGDCONT=1,\"IP\",\"" CONFIG_MODEM_GSM_APN "\""),
+	/* connect to network */
+	SETUP_CMD_NOHANDLE("ATD*99#")
+};
+
+static void gsm_configure(struct k_work *work)
+{
+	int r = -1;
+	struct gsm_modem *gsm = CONTAINER_OF(work, struct gsm_modem,
+					       gsm_configure_work);
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c]: indentation issue

File: samples/net/gsm_modem/prj.conf (hunk 8 above)
Comment (pfl) [reply samples/net/gsm_modem/prj.conf]: Done

File: samples/net/gsm_modem/src/main.c (hunk 9 above)
Comment (pfl) [reply samples/net/gsm_modem/src/main.c]: Fixed.

File: drivers/modem/gsm_ppp.c (hunk 10 above)
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: Yes.

File: drivers/modem/gsm_ppp.c (hunk 12 above)
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: Yes.

File: drivers/modem/gsm_ppp.c (hunk 13 above)
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: Done.

File: drivers/modem/gsm_ppp.c (hunk 14 above)
Comment (pfl) [reply drivers/modem/gsm_ppp.c]: Fixed.

**Review by jukkar — APPROVED** [review]

**Pull request merged**